test/test_server.py

```diff
@@ -316,6 +317,14 @@ class TestHTTPS(tservers.HTTPProxTest, CommonMixin, TcpMixin):
     clientcerts = True
 
     def test_clientcert(self):
+        self.config.clientcerts = os.path.join(
+            tutils.test_data.path("data/clientcert"), "client.pem")
+        f = self.pathod("304")
+        assert f.status_code == 304
+        assert self.server.last_log()["request"]["clientcert"]["keyinfo"]
+
+    def test_clientcerts(self):
+        self.config.clientcerts = tutils.test_data.path("data/clientcert")
```

Comment:
You have obviously noticed that I changed the default behavior of the assignment of clientcerts from the setup of ProxTestBase class to the actual test function themselves. I did this for important reasons, that I would like you to analyze and assess.

Since the clientcerts assignment was made at object instantiation, the only way to have changed the setting would be to add additional logic to get_proxy_config(). I did this at first -- ProxTestBase's `clientcerts` class variable able to handle the numbers 0, 1 and 2 as a reference to whether the config should use None, file, or directory. The problem here is that the config is only made at object instantiation. This means another TestHTTPS class must be made with a different clientcert variable value. But, this also meant a re-running of all of the tests that belong to the classes that TestHTTPS inherits from (tservers.HTTPProxtTest, CommonMixin, TcpMixin).

This is what led me to investigate if I could manually change the real clientcerts proxy config from the test itself effectively -- No need to recreate objects just to change the clientcerts configuration.

The reason why I knew (or was convinced) that I could set the clientcerts config manually during the tests is because the class ProxyConfig `__init__` does not perform any special duties with the clientcerts dict argument returned from get_proxy_config() that I could not do outside of the object instantiation phase. A simple search for "clientcerts" in the project will show that it is not widely used and may be safe to control outside of object instantiation.


see https://github.com/mitmproxy/mitmproxy/blob/a978c6b9cee134d01f84216f3fbcf1ae9303d965/test/tservers.py#L127

According to your recommended removal of `clientcerts = True` from class TestHTTPS, it may be recommended to also have me remove the use of `clientcerts` in ProxTestBase. I'm referring to the class variable and the dict entry altogether. This would cause all proxies to use no clientcerts unless specifically configured as done in the new tests that are shown here.

Of course, this approach can only continue to work as long as the clientcerts setting can be so freely modified during the server's existence.

I've checked to make sure this logic is not used elsewhere. So, if you agree, I'll also remove this cruft from the ProxTestBase.